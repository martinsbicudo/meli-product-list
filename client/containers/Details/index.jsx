import React, { useState, useEffect } from 'react';
import { string, func } from 'prop-types';
import ProductDetails from '@Containers/ProductDetails';
import Loading from '@Components/Loading';
import EmptyAlert from '@Containers/EmptyAlert';
import Link from 'next/link';
import { useLoadingStatus } from '@Hooks';
import { getProductDetails } from '@Actions';
import style from './style.scss';

const Details = ({
  id,
  setTitle,
}) => {
  const [product, setProduct] = useState(false);

  const [{
    loading,
    status,
  }, {
    setLoading,
    setStatus,
  }] = useLoadingStatus();

  useEffect(() => {
    setLoading(true);

    getProductDetails({
      id,
      resolve({ data }) {
        setStatus({
          success: true,
          error: false,
        });
        setProduct(data);
        setTitle(`${data.title} no Mercado Livre`);
      },
      reject() {
        setStatus({
          success: false,
          error: true,
        });
        setTitle('Produto não encontrado no Mercado Livre');
      },
      finish() {
        setLoading(false);
      },
    });
  }, []);

  const isLoading = (conditional = true) => loading
    || (!status.success && !status.error && conditional);

  return (
    <section className={style.Details}>
      <Loading isLoading={isLoading()}>
        {() => (
          <>
            {!isLoading() && !product && (
              <EmptyAlert
                empty={{
                  title: 'Produto não encontrado',
                  description: (
                    <>
                      {'Volte para a home clicando '}
                      <Link href="/">
                        <a href="javascript();">
                          aqui
                        </a>
                      </Link>
                    </>
                  ),
                }}
              />
            )}
            {!isLoading() && !status.error && product && (
              <ProductDetails product={product} />
            )}
          </>
        )}
      </Loading>
    </section>
  );
};

Details.propTypes = {
  id: string.isRequired,
  setTitle: func.isRequired,
};

export default Details;
