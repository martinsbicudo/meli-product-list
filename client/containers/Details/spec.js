import Details from '.';
import { useLoadingStatus } from '@Hooks';
import { getProductDetails } from '@Actions';

// MOCK USE-LOADING-STATUS
const initialState = [{
  loading: false,
  status: {
    success: false,
    error: false,
  },
}, {
  setLoading: jest.fn(),
  setStatus: jest.fn(),
}];
jest.mock('../../hooks', () => ({
  useLoadingStatus: jest.fn(),
}));
// /

// MOCK USE-LOADING-STATUS
const testMessageTitle = 'TEST_TITLE';
const testMessagePrice = '300';
const testMessageImage = 'TEST_IMAGE';
const testMessageDescription = 'TEST_DESCRIPTION';
const initialProducts = ({
  resolve,
  reject,
  finish,
}) => {
  resolve({
    data: {
      id: Math.random(),
      title: testMessageTitle,
      price: testMessagePrice,
      pictures: [{
        url: testMessageImage,
      }],
      description: testMessageDescription,
    },
  });
  reject();
  finish();
};
jest.mock('../../actions', () => ({
  getProductDetails: jest.fn(),
}));
// /

describe('Containers - <Details />', () => {
  const id = 'TEST_ID';
  const createWrapper = () => render(
    <Details search={id} setTitle={jest.fn()} />,
  );

  test('Should finds 5 LoadingBar', () => {
    useLoadingStatus.mockImplementation(() => initialState);
    const { container } = createWrapper();

    expect(container.querySelectorAll('.LoadingBar')).toHaveLength(5);
  });

  test('Should finds EmptyAlert', () => {
    useLoadingStatus.mockImplementation(() => ([{
      ...initialState[0],
      status: {
        success: true,
        error: false,
      },
    }, { ...initialState[1] }]));
    const { getByText } = createWrapper();

    expect(getByText('Produto nÃ£o encontrado')).toBeInTheDocument();
  });

  test('Should finds Product Description', () => {
    useLoadingStatus.mockImplementation(() => ([{
      ...initialState[0],
      status: {
        success: true,
        error: false,
      },
    }, { ...initialState[1] }]));
    getProductDetails.mockImplementation(initialProducts);
    const { getByText } = createWrapper();

    expect(getByText(testMessageDescription)).toBeInTheDocument();
  });

  test('Wrapper Snapshot', () => {
    useLoadingStatus.mockImplementation(() => initialState);
    const { asFragment } = createWrapper();

    expect(asFragment()).toMatchSnapshot();
  });
});
