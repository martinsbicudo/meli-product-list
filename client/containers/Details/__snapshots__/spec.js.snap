// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers - <Details /> Wrapper Snapshot 1`] = `
<DocumentFragment>
  <section
    class="Details"
  >
    <div
      class="Loading"
    >
      <div
        class="LoadingBars"
      >
        <i
          class="LoadingBar"
        />
        <i
          class="LoadingBar"
        />
        <i
          class="LoadingBar"
        />
        <i
          class="LoadingBar"
        />
        <i
          class="LoadingBar"
        />
      </div>
    </div>
  </section>
</DocumentFragment>
`;
